mesonbuild/mlog.py

```diff
@@ -369,11 +376,11 @@ def exception(self, e: Exception, prefix: T.Optional[AnsiDecorator] = None) -> N
             prefix = red('ERROR:')
         self.log()
         args = []  # type: T.List[T.Union[AnsiDecorator, str]]
-        if all(getattr(e, a, None) is not None for a in ['file', 'lineno', 'colno']):
-            # Mypy doesn't follow hasattr, and it's pretty easy to visually inspect
-            # that this is correct, so we'll just ignore it.
-            path = get_relative_path(Path(e.file), Path(os.getcwd()))  # type: ignore
-            args.append(f'{path}:{e.lineno}:{e.colno}:')  # type: ignore
+
+        from .mesonlib import MesonException
+        if isinstance(e, MesonException) and e.file is not None:
+            path = get_relative_path(Path(e.file), Path(os.getcwd()))
+            args.append(f'{path}:{or_else(e.lineno, -1)}:{or_else(e.colno, -1)}:')
```

Comment:
> The reality is that these checks are looking for a MesonException, let's
just actually check for a MesonException, and make the code clearer, and
mypy happier.

Interpreterbase actually set that on all exceptions: https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase/interpreterbase.py#L183. I agree it only make sense to do that on MesonException because otherwise we print the full backtrace anyway. Maybe we should only catch MesonException there.

We do similar getattr() there too: https://github.com/mesonbuild/meson/blob/master/mesonbuild/mesonmain.py#L61.

Not sure we should fallback to -1, in practice if e.file is not None, then e.lineno and e.colno cannot be None. Maybe just assert they are not None?

Reply:
> Interpreterbase actually set that on all exceptions: https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase/interpreterbase.py#L183. 

We should really be using `from caught_exception raise MesonException` in that case... Plus I see some cases in there that would benefit from the `from_node` method...

I'd have to look some more to see if we really would never hit a case that's true. I have this vague feeling that cmake can do that in some cases, but that may not be true

Reply:
I think we should put that change on the side for now. It is not related to the goal of the PR and we probably could do more cleanup around the exception handling in its own PR.

Currently any exception other than MesonException will print backtrace followed by `meson.build:3:0: ERROR: Unhandled python exception`, which is useful info to keep.

Reply:
It looks like mlog makes the assumption you're proposing, so I guess it's safe to do here too.

Reply:
Agreed, I'll send a separate PR that just mucks around with the error handling. There's several things we can do there that I think are nicer, and will still get the same result.

Reply:
IMHO the error handling is correct as it currently is.
- We must never raise MesonException for internal bugs.
- We want to smuggle line numbers out of *any* exception.
- It seems off to raise a new exception here, since under MESON_FORCE_BACKTRACE=1 we will then return multiple chained exceptions when what we really wanted is to just get the original exception.

Reply:
Adding random attributes to things is always a code smell. I have a branch, but I'm running into weird issues in the way mparser sets things up. I'll send that out separately and we can discuss how to handle that there.

Reply:
I've gone ahead and dropped this patch for now, we can discuss changing error handling separately.